File: app/src/main/kotlin/com/example/mpesatracker/parser/MpesaParser.kt
package com.example.mpesatracker.parser

import com.example.mpesatracker.data.model.MpesaTransaction
import java.text.SimpleDateFormat
import java.util.*

object MpesaParser {

    enum class TransactionType {
        SENT,
        RECEIVED,
        PAYBILL,
        BUY_GOODS,
        WITHDRAW,
        DEPOSIT,
        AIRTIME,
        FULIZA_BORROW,
        FULIZA_REPAY,
        REVERSAL,
        UNKNOWN
    }

    private fun extractAmount(msg: String): Double? {
        val amountRegex = """(?:Ksh|KSh|KES)\s*([\d,]+(?:\.\d{1,2})?)""".toRegex(RegexOption.IGNORE_CASE)
        val match = amountRegex.find(msg) ?: return null
        return match.groupValues[1].replace(",", "").toDoubleOrNull()
    }

    private fun extractConfirmationCode(msg: String): String? {
        val codeRegex = """\b([A-Z0-9]{10})\b""".toRegex()
        return codeRegex.find(msg)?.groupValues?.get(1)
    }

    private fun extractDate(msg: String): Long {
        val dateTimeRegex = """on\s+(\d{1,2})/(\d{1,2})/(\d{2,4})\s+at\s+(\d{1,2}):(\d{2})\s+(AM|PM)""".toRegex(RegexOption.IGNORE_CASE)
        val match = dateTimeRegex.find(msg)
        if (match != null) {
            try {
                val day = match.groupValues[1].toInt()
                val month = match.groupValues[2].toInt() - 1
                var year = match.groupValues[3].toInt()
                if (year < 100) year += 2000
                
                var hour = match.groupValues[4].toInt()
                val minute = match.groupValues[5].toInt()
                val amPm = match.groupValues[6].uppercase()
                
                if (amPm == "PM" && hour < 12) {
                    hour += 12
                } else if (amPm == "AM" && hour == 12) {
                    hour = 0
                }
                
                val calendar = Calendar.getInstance()
                calendar.set(year, month, day, hour, minute, 0)
                calendar.set(Calendar.MILLISECOND, 0)
                return calendar.timeInMillis
            } catch (e: Exception) {
                // fallback to basic date matching
            }
        }
        
        val dateRegex = """(\d{1,2})/(\d{1,2})/(\d{2,4})""".toRegex()
        val dateMatch = dateRegex.find(msg)
        if (dateMatch != null) {
            try {
                val day = dateMatch.groupValues[1].toInt()
                val month = dateMatch.groupValues[2].toInt() - 1
                var year = dateMatch.groupValues[3].toInt()
                if (year < 100) year += 2000
                val calendar = Calendar.getInstance()
                calendar.set(year, month, day, 12, 0, 0)
                return calendar.timeInMillis
            } catch (e: Exception) {
                // fallback
            }
        }
        return System.currentTimeMillis()
    }

    private fun extractParty(msg: String, afterKeyword: String): String? {
        val lowerMsg = msg.lowercase()
        val idx = lowerMsg.indexOf(afterKeyword.lowercase())
        if (idx == -1) return null
        val rest = msg.substring(idx + afterKeyword.length).trim()
        val endRegex = """\b(\d{10,}|on\s|\bfor\b|\bat\b)""".toRegex(RegexOption.IGNORE_CASE)
        val match = endRegex.find(rest)
        return if (match != null) {
            rest.substring(0, match.range.first).trim()
        } else {
            if (rest.length > 40) rest.substring(0, 40).trim() else rest
        }
    }

    private fun suggestCategory(type: TransactionType, msg: String): String? {
        val lower = msg.lowercase()
        if (type == TransactionType.FULIZA_BORROW || type == TransactionType.FULIZA_REPAY) {
            return "LOANS"
        }
        if (type == TransactionType.AIRTIME) {
            return "INTERNET"
        }
        if (lower.contains("kplc") || lower.contains("kenya power") || lower.contains("electricity") || lower.contains("power")) {
            return "ELECTRICITY"
        }
        if (lower.contains("safaricom") && (lower.contains("data") || lower.contains("bundle") || lower.contains("internet"))) {
            return "INTERNET"
        }
        if (lower.contains("nairobi water") || lower.contains("water")) {
            return "ELECTRICITY"
        }
        if (lower.contains("rent") || lower.contains("landlord")) {
            return "RENT"
        }
        return null
    }

    fun isMpesaMessage(body: String?, sender: String = ""): Boolean {
        if (body.isNullOrBlank()) return false
        if (sender.lowercase().contains("mpesa") || sender.lowercase().contains("m-pesa")) return true
        val upperPrefix = body.trim().take(20).uppercase()
        if (upperPrefix.contains("MPESA") || upperPrefix.contains("M-PESA")) return true
        
        // Starts with 10-char alphanumeric code + space
        val codeStartRegex = """^[A-Z0-9]{10}\s""".toRegex(RegexOption.IGNORE_CASE)
        if (codeStartRegex.containsMatchIn(body.trim())) {
            val amountRegex = """(?:Ksh|KSh|KES)\s*[\d,]+""".toRegex(RegexOption.IGNORE_CASE)
            return amountRegex.containsMatchIn(body)
        }
        return false
    }

    fun parseMpesaSms(body: String?, sender: String = "", defaultDate: Long = System.currentTimeMillis()): MpesaTransaction? {
        if (body.isNullOrBlank()) return null

        val isMpesa = isMpesaMessage(body, sender)
        if (!isMpesa) {
            // Last resort keyword check
            val keywordRegex = """\b(m-pesa|mpesa|fuliza)\b""".toRegex(RegexOption.IGNORE_CASE)
            if (!keywordRegex.containsMatchIn(body)) return null
        }

        val msg = body.trim()
        val lower = msg.lowercase()
        val amount = extractAmount(msg) ?: return null
        val code = extractConfirmationCode(msg)
        val txDate = extractDate(msg)

        var type = TransactionType.UNKNOWN
        var party: String? = null
        var isExpense = true
        var isIncome = false
        var balance: Double? = null
        var accountNumber: String? = null
        var paybillNumber: String? = null

        // Extract balance
        val balanceRegex = """(?:new\s+m-pesa\s+balance|balance\s+is|balance:)\s*(?:Ksh|KSh|KES)?\s*([\d,]+(?:\.\d{1,2})?)""".toRegex(RegexOption.IGNORE_CASE)
        val balanceMatch = balanceRegex.find(msg)
        if (balanceMatch != null) {
            balance = balanceMatch.groupValues[1].replace(",", "").toDoubleOrNull()
        }

        if (lower.contains("fuliza")) {
            val borrowRegex = """you have borrowed|you borrowed|fuliza loan of""".toRegex()
            val repayRegex = """repaid|repayment|recovered from your m-pesa""".toRegex()
            if (borrowRegex.containsMatchIn(lower)) {
                type = TransactionType.FULIZA_BORROW
                isExpense = true
            } else if (repayRegex.containsMatchIn(lower)) {
                type = TransactionType.FULIZA_REPAY
                isExpense = true
            } else {
                type = TransactionType.FULIZA_BORROW
                isExpense = true
            }
        } else if ("""confirmed[.,]?\s+you have received|you have received|received\s+(?:Ksh|KSh)""".toRegex().containsMatchIn(lower)) {
            type = TransactionType.RECEIVED
            party = extractParty(msg, "from")
            isExpense = false
            isIncome = true
        } else if ("""deposited to m-pesa|deposit of""".toRegex().containsMatchIn(lower)) {
            type = TransactionType.DEPOSIT
            party = extractParty(msg, "by")
            isExpense = false
            isIncome = true
        } else if (lower.contains("reversal") || lower.contains("reversed")) {
            type = TransactionType.REVERSAL
            isExpense = false
            isIncome = true
        } else if ("""airtime purchase|airtime of|bought airtime""".toRegex().containsMatchIn(lower)) {
            type = TransactionType.AIRTIME
            isExpense = true
        } else if ("""confirmed[.,]?\s+(?:Ksh|KSh).*paid to|paid to|paybill|for account""".toRegex().containsMatchIn(lower)) {
            type = TransactionType.PAYBILL
            party = extractParty(msg, "paid to")
            isExpense = true

            // Account number
            val acRegex = """account\s+([A-Z0-9\-]+)""".toRegex(RegexOption.IGNORE_CASE)
            accountNumber = acRegex.find(msg)?.groupValues?.get(1)

            // Paybill number
            val pbRegex = """(?:paybill number|business number)\s*:?\s*(\d+)""".toRegex(RegexOption.IGNORE_CASE)
            paybillNumber = pbRegex.find(msg)?.groupValues?.get(1)
        } else if (("""paid to|buy goods|merchant""".toRegex().containsMatchIn(lower)) && !lower.contains("for account")) {
            type = TransactionType.BUY_GOODS
            party = extractParty(msg, "to")
            isExpense = true
        } else if (lower.contains("withdrawn from") || lower.contains("cash out")) {
            type = TransactionType.WITHDRAW
            party = extractParty(msg, "from")
            isExpense = true
        } else if ("""confirmed[.,]?\s+(?:Ksh|KSh).*sent to|sent to|you sent""".toRegex().containsMatchIn(lower)) {
            type = TransactionType.SENT
            party = extractParty(msg, "to")
            isExpense = true
        } else {
            type = TransactionType.UNKNOWN
            isExpense = true
        }

        val suggestedCategory = suggestCategory(type, msg)

        return MpesaTransaction(
            id = code ?: "${txDate}-${(0..10000).random()}",
            confirmationCode = code,
            type = type.name,
            amount = amount,
            party = party,
            accountNumber = accountNumber,
            paybillNumber = paybillNumber,
            date = txDate,
            rawMessage = msg,
            isExpense = isExpense,
            isIncome = isIncome,
            balance = balance,
            category = suggestedCategory,
            categorized = suggestedCategory != null,
            incomeSource = if (isIncome) null else null, // income source starts null
            note = null,
            createdAt = defaultDate
        )
    }
}

File: src/utils/mpesaParser.js
/**
 * M-Pesa SMS message parser
 * Handles all types of M-Pesa transaction messages
 */

export const TRANSACTION_TYPES = {
  SENT: 'SENT',
  RECEIVED: 'RECEIVED',
  PAYBILL: 'PAYBILL',
  BUY_GOODS: 'BUY_GOODS',
  WITHDRAW: 'WITHDRAW',
  DEPOSIT: 'DEPOSIT',
  AIRTIME: 'AIRTIME',
  FULIZA_BORROW: 'FULIZA_BORROW',
  FULIZA_REPAY: 'FULIZA_REPAY',
  REVERSAL: 'REVERSAL',
  UNKNOWN: 'UNKNOWN',
};

/**
 * Extract amount from M-Pesa message.
 * Handles formats like: Ksh1,234.00 | KSh200 | KES 1,000.50
 */
function extractAmount(msg) {
  const match = msg.match(/(?:Ksh|KSh|KES)\s*([\d,]+(?:\.\d{1,2})?)/i);
  if (!match) return null;
  return parseFloat(match[1].replace(/,/g, ''));
}

/**
 * Extract M-Pesa confirmation code (e.g. QGH2K345RT)
 */
function extractConfirmationCode(msg) {
  const match = msg.match(/\b([A-Z0-9]{10})\b/);
  return match ? match[1] : null;
}

/**
 * Extract a date string from M-Pesa message.
 * Handles patterns like "22/2/26" or "22/02/2026" or "22 Feb 2026"
 */
function extractDate(msg) {
  // Try DD/M/YY or DD/MM/YYYY
  let match = msg.match(/(\d{1,2}\/\d{1,2}\/\d{2,4})/);
  if (match) {
    const parts = match[1].split('/');
    const day = parseInt(parts[0], 10);
    const month = parseInt(parts[1], 10) - 1;
    let year = parseInt(parts[2], 10);
    if (year < 100) year += 2000;
    const d = new Date(year, month, day);
    return isNaN(d.getTime()) ? new Date() : d;
  }
  // Fallback
  return new Date();
}

/**
 * Extract the recipient or sender name from message.
 */
function extractParty(msg, afterKeyword) {
  const idx = msg.toLowerCase().indexOf(afterKeyword.toLowerCase());
  if (idx === -1) return null;
  const rest = msg.slice(idx + afterKeyword.length).trim();
  // Name is everything up to a phone number, "on", or "for"
  const end = rest.search(/\b(\d{10,}|on\s|\bfor\b|\bat\b)/i);
  if (end === -1) return rest.slice(0, 40).trim();
  return rest.slice(0, end).trim();
}

/**
 * Determine the suggested category based on transaction type and content.
 */
function suggestCategory(type, msg) {
  const lower = msg.toLowerCase();

  if (type === TRANSACTION_TYPES.FULIZA_BORROW || type === TRANSACTION_TYPES.FULIZA_REPAY) {
    return 'LOANS';
  }
  if (type === TRANSACTION_TYPES.AIRTIME) {
    return 'INTERNET';
  }
  if (lower.includes('kplc') || lower.includes('kenya power') || lower.includes('electricity') || lower.includes('power')) {
    return 'ELECTRICITY';
  }
  if (lower.includes('safaricom') && (lower.includes('data') || lower.includes('bundle') || lower.includes('internet'))) {
    return 'INTERNET';
  }
  if (lower.includes('nairobi water') || lower.includes('water')) {
    return 'ELECTRICITY';
  }
  if (lower.includes('rent') || lower.includes('landlord')) {
    return 'RENT';
  }
  if (type === TRANSACTION_TYPES.RECEIVED || type === TRANSACTION_TYPES.DEPOSIT) {
    return null; // income — no category needed
  }

  return null; // to be determined by user
}

/**
 * Main parser: takes a raw SMS body and sender, returns a parsed transaction object.
 * Returns null if the message is not a recognised M-Pesa transaction.
 */
export function parseMpesaSms(body, sender = '', date = new Date()) {
  if (!body) return null;

  // Only process messages from Mpesa / MPESA sender
  const isMpesaSender = /mpesa|m-pesa/i.test(sender) || /mpesa|m-pesa/i.test(body.slice(0, 20));

  // Quick confirmation code check — M-Pesa messages always start with a code
  const hasCode = /^[A-Z0-9]{10}\s/i.test(body.trim());

  if (!isMpesaSender && !hasCode) {
    // Last resort: check for common M-Pesa keywords
    const hasMpesaKeyword = /\b(m-pesa|mpesa|fuliza)\b/i.test(body);
    if (!hasMpesaKeyword) return null;
  }

  const msg = body.trim();
  const lower = msg.toLowerCase();
  const amount = extractAmount(msg);
  const code = extractConfirmationCode(msg);
  const txDate = extractDate(msg);

  let type = TRANSACTION_TYPES.UNKNOWN;
  let party = null;
  let isExpense = true;
  let isIncome = false;
  let balance = null;
  let accountNumber = null;
  let paybillNumber = null;

  // Extract new balance
  const balanceMatch = msg.match(/(?:new\s+m-pesa\s+balance|balance\s+is|balance:)\s*(?:Ksh|KSh|KES)?\s*([\d,]+(?:\.\d{1,2})?)/i);
  if (balanceMatch) {
    balance = parseFloat(balanceMatch[1].replace(/,/g, ''));
  }

  // ─── FULIZA ──────────────────────────────────────────────────────────────
  if (/fuliza/i.test(lower)) {
    if (/you have borrowed|you borrowed|fuliza loan of/i.test(lower)) {
      type = TRANSACTION_TYPES.FULIZA_BORROW;
      isExpense = true;
    } else if (/repaid|repayment|recovered from your m-pesa/i.test(lower)) {
      type = TRANSACTION_TYPES.FULIZA_REPAY;
      isExpense = true;
    } else {
      type = TRANSACTION_TYPES.FULIZA_BORROW;
      isExpense = true;
    }
  }
  // ─── RECEIVED ────────────────────────────────────────────────────────────
  else if (/confirmed[.,]?\s+you have received|you have received|received\s+(?:Ksh|KSh)/i.test(lower)) {
    type = TRANSACTION_TYPES.RECEIVED;
    party = extractParty(msg, 'from');
    isExpense = false;
    isIncome = true;
  }
  // ─── DEPOSIT ─────────────────────────────────────────────────────────────
  else if (/deposited to m-pesa|deposit of/i.test(lower)) {
    type = TRANSACTION_TYPES.DEPOSIT;
    party = extractParty(msg, 'by');
    isExpense = false;
    isIncome = true;
  }
  // ─── REVERSAL ────────────────────────────────────────────────────────────
  else if (/reversal|reversed/i.test(lower)) {
    type = TRANSACTION_TYPES.REVERSAL;
    isExpense = false;
    isIncome = true;
  }
  // ─── AIRTIME ─────────────────────────────────────────────────────────────
  else if (/airtime purchase|airtime of|bought airtime/i.test(lower)) {
    type = TRANSACTION_TYPES.AIRTIME;
    isExpense = true;
  }
  // ─── PAYBILL ─────────────────────────────────────────────────────────────
  else if (/confirmed[.,]?\s+(?:Ksh|KSh).*paid to|paid to|paybill|for account/i.test(lower)) {
    type = TRANSACTION_TYPES.PAYBILL;
    party = extractParty(msg, 'paid to');
    isExpense = true;
    // Extract account number
    const acMatch = msg.match(/account\s+([A-Z0-9\-]+)/i);
    if (acMatch) accountNumber = acMatch[1];
    // Extract paybill number
    const pbMatch = msg.match(/(?:paybill number|business number)\s*:?\s*(\d+)/i);
    if (pbMatch) paybillNumber = pbMatch[1];
  }
  // ─── BUY GOODS ───────────────────────────────────────────────────────────
  else if (/paid to|buy goods|merchant/i.test(lower) && !/for account/i.test(lower)) {
    type = TRANSACTION_TYPES.BUY_GOODS;
    party = extractParty(msg, 'to');
    isExpense = true;
  }
  // ─── WITHDRAW ────────────────────────────────────────────────────────────
  else if (/withdrawn from|cash out/i.test(lower)) {
    type = TRANSACTION_TYPES.WITHDRAW;
    party = extractParty(msg, 'from');
    isExpense = true;
  }
  // ─── SENT ────────────────────────────────────────────────────────────────
  else if (/confirmed[.,]?\s+(?:Ksh|KSh).*sent to|sent to|you sent/i.test(lower)) {
    type = TRANSACTION_TYPES.SENT;
    party = extractParty(msg, 'to');
    isExpense = true;
  }
  // ─── FALLBACK: amount present means expense ───────────────────────────────
  else if (amount !== null) {
    type = TRANSACTION_TYPES.UNKNOWN;
    isExpense = true;
  } else {
    return null;
  }

  if (amount === null) return null;

  const suggestedCategory = suggestCategory(type, msg);

  return {
    id: code || `${Date.now()}-${Math.random().toString(36).slice(2)}`,
    confirmationCode: code,
    type,
    amount,
    party,
    accountNumber,
    paybillNumber,
    date: txDate,
    rawMessage: msg,
    isExpense,
    isIncome,
    balance,
    // Expense categorisation
    category: suggestedCategory,
    categorized: suggestedCategory !== null,
    // Income source categorisation (null until user selects)
    incomeSource: isIncome ? null : undefined,
    createdAt: date,
  };
}

/**
 * Check if a given SMS body looks like an M-Pesa message at all.
 */
export function isMpesaMessage(body, sender = '') {
  if (!body) return false;
  if (/mpesa|m-pesa/i.test(sender)) return true;
  if (/mpesa|m-pesa|fuliza/i.test(body)) return true;
  // Starts with 10-char alphanumeric code + space
  if (/^[A-Z0-9]{10}\s/i.test(body.trim())) {
    const hasAmount = /(?:Ksh|KSh|KES)\s*[\d,]+/i.test(body);
    return hasAmount;
  }
  return false;
}
